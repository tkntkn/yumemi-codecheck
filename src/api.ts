async function fetchResas(resource: string) {
  const response = await fetch(
    `https://opendata.resas-portal.go.jp${resource}`,
    {
      headers: {
        "X-API-KEY": RESAS_API_KEY,
      },
    }
  );
  return response.json();
}

export type Prefecture = {
  prefCode: number;
  prefName: string;
};

let prefecturesCache: Prefecture[];

export async function fetchPrefectures(): Promise<Prefecture[]> {
  if (!prefecturesCache) {
    const response = (await fetchResas("/api/v1/prefectures")) as any;
    const prefectures = response.result;
    prefecturesCache = prefectures;
  }

  return prefecturesCache;
}

type Population = number[];

const populationCache: { [prefCode in number]: Population } = {};

export async function fetchPopulation(prefCode: number): Promise<Population> {
  if (!populationCache[prefCode]) {
    const response = (await fetchResas(
      `/api/v1/population/composition/perYear?prefCode=${prefCode}&cityCode=-`
    )) as any;
    const population = response.result.data
      .find(({ label }) => label === "総人口")
      .data.map((data) => data.value);
    populationCache[prefCode] = population;
  }

  return populationCache[prefCode];
}
