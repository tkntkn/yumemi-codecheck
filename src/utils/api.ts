async function fetchResas(resource: string) {
  const response = await fetch(
    `https://opendata.resas-portal.go.jp${resource}`,
    {
      headers: {
        "X-API-KEY": RESAS_API_KEY,
      },
    }
  );
  return response.json();
}

export type Prefecture = {
  prefCode: number;
  prefName: string;
};
let prefecturesCache: Prefecture[];
export async function fetchPrefectures(): Promise<Prefecture[]> {
  if (!prefecturesCache) {
    const response = (await fetchResas("/api/v1/prefectures")) as {
      result: Prefecture[];
    };
    const prefectures = response.result;
    prefecturesCache = prefectures;
  }

  return prefecturesCache;
}

type Population = number[];
const populationCache: { [prefCode in number]: Population } = {};
export async function fetchPopulation(prefCode: number): Promise<Population> {
  if (!populationCache[prefCode]) {
    const response = (await fetchResas(
      `/api/v1/population/composition/perYear?prefCode=${prefCode}&cityCode=-`
    )) as {
      result: {
        data: Array<{
          label: string;
          data: Array<{
            year: number;
            value: number;
          }>;
        }>;
      };
    };
    // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
    const population = response.result.data
      .find((data) => data.label === "総人口")!
      .data.map((data) => data.value);
    populationCache[prefCode] = population;
  }

  return populationCache[prefCode];
}
