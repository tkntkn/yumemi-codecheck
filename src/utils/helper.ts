import { fetchPopulation, Prefecture } from "./api";

export type Population = {
  code: number;
  name: string;
  population: number[];
};

export async function constructPopulation(
  code: number,
  prefs: Prefecture[]
): Promise<Population> {
  return {
    code,
    // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
    name: prefs.find((pref) => pref.prefCode === code)!.prefName,
    population: await fetchPopulation(code),
  };
}
