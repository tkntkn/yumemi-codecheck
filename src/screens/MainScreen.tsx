import { ReactElement, useState, useCallback } from "react";
import useSWR from "swr";
import { fetchPrefectures } from "../utils/api";
import { CheckboxList } from "../components/CheckboxList";
import { constructPopulation, Population } from "../utils/helper";
import { ResasPopulationTransitionChart } from "../components/ResasPopulationTransitionChart";

export function MainScreen(): ReactElement {
  const { data: prefs, error: prefsError } = useSWR("PREFS", fetchPrefectures);

  const [populations, setPopulations] = useState<Population[]>([]);

  const onPrefChecked = useCallback(
    async (key: string) => {
      try {
        // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
        const population = await constructPopulation(+key, prefs!);
        setPopulations((pops) => pops.concat(population));
      } catch {
        // TODO: アラートUIを作る
        alert("人口データの取得に失敗しました。");
      }
    },
    [prefs]
  );

  const onPrefUnchecked = useCallback((key: string) => {
    setPopulations((pops) => pops.filter((pop) => pop.code !== +key));
  }, []);

  return (
    <div className="MainScreen">
      <header className="MainScreen__header">
        <h1 className="MainScreen__headerTitle">人口推移ビューワー</h1>
      </header>
      {prefsError ? (
        <div>都道府県データの取得に失敗しました。</div>
      ) : prefs ? (
        <>
          <div className="MainScreen__prefs">
            <h2 className="MainScreen__prefsTitle">都道府県</h2>
            <CheckboxList
              items={prefs.map(({ prefCode, prefName }) => ({
                key: prefCode.toString(),
                label: prefName,
              }))}
              onCheck={onPrefChecked}
              onUncheck={onPrefUnchecked}
            />
          </div>
          <div className="MainScreen__chart">
            <ResasPopulationTransitionChart
              series={populations.map(({ name, code, population }) => ({
                type: "line",
                id: code.toString(),
                name: name,
                data: population,
              }))}
            />
          </div>
        </>
      ) : (
        <div>都道府県データの取得中...</div>
      )}
    </div>
  );
}
