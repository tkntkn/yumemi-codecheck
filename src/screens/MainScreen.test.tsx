import {
  cleanup,
  render,
  screen,
  act,
  fireEvent,
} from "@testing-library/react";
import { SWRConfig } from "swr";
import { MainScreen } from "./MainScreen";
import * as api from "../utils/api";

describe("MainScreen", () => {
  afterEach(() => {
    cleanup();
  });

  test("render", async () => {
    jest.spyOn(api, "fetchPrefectures").mockImplementation(async () => [
      {
        prefCode: 1,
        prefName: "東京都",
      },
      {
        prefCode: 2,
        prefName: "大阪府",
      },
    ]);

    jest
      .spyOn(api, "fetchPopulation")
      .mockImplementation(async () => [10, 20, 30, 40, 50]);

    let asFragment: () => DocumentFragment;
    const action = act(async () => {
      asFragment = render(
        <SWRConfig value={{ provider: () => new Map() }}>
          <MainScreen />
        </SWRConfig>
      ).asFragment;
    });
    const title = screen.getByText("人口推移ビューワー");
    expect(title).toBeInTheDocument();
    const loadMessage = screen.getByText("都道府県データの取得中...");
    expect(loadMessage).toBeInTheDocument();
    await action;
    const tokyo = screen.getByText("東京都");
    expect(tokyo).toBeInTheDocument();
    await act(async () => void fireEvent.click(tokyo));
    // eslint-disable-next-line @typescript-eslint/ban-ts-comment
    // @ts-ignore
    expect(asFragment()).toMatchSnapshot();
  });

  test("render fetchPrefectures error", async () => {
    jest.spyOn(api, "fetchPrefectures").mockImplementation(async () => {
      throw "error mock";
    });

    const action = act(
      async () =>
        void render(
          <SWRConfig value={{ provider: () => new Map() }}>
            <MainScreen />
          </SWRConfig>
        )
    );
    const title = screen.getByText("人口推移ビューワー");
    expect(title).toBeInTheDocument();
    const loadMessage = screen.getByText("都道府県データの取得中...");
    expect(loadMessage).toBeInTheDocument();
    await action;
    const failMessage =
      screen.getByText("都道府県データの取得に失敗しました。");
    expect(failMessage).toBeInTheDocument();
  });

  test("alert fetchPopulation error", async () => {
    jest.spyOn(api, "fetchPrefectures").mockImplementation(async () => [
      {
        prefCode: 1,
        prefName: "東京都",
      },
      {
        prefCode: 2,
        prefName: "大阪府",
      },
    ]);

    jest.spyOn(api, "fetchPopulation").mockImplementation(async () => {
      throw "error mock";
    });

    const alertMock = jest
      .spyOn(window, "alert")
      .mockImplementation(() => void 0);

    const action = act(
      async () =>
        void render(
          <SWRConfig value={{ provider: () => new Map() }}>
            <MainScreen />
          </SWRConfig>
        )
    );
    const title = screen.getByText("人口推移ビューワー");
    expect(title).toBeInTheDocument();
    const loadMessage = screen.getByText("都道府県データの取得中...");
    expect(loadMessage).toBeInTheDocument();
    await action;
    const tokyo = screen.getByText("東京都");
    expect(tokyo).toBeInTheDocument();
    await act(async () => void fireEvent.click(tokyo));
    expect(alertMock).toHaveBeenCalledWith("人口データの取得に失敗しました。");
  });
});
