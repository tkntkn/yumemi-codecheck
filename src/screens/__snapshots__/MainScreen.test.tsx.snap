// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MainScreen render 1`] = `
<DocumentFragment>
  <div
    class="MainScreen"
  >
    <header
      class="MainScreen__header"
    >
      <h1
        class="MainScreen__headerTitle"
      >
        人口推移ビューワー
      </h1>
    </header>
    <div
      class="MainScreen__prefs"
    >
      <h2
        class="MainScreen__prefsTitle"
      >
        都道府県
      </h2>
      <div
        class="CheckboxList"
      >
        <label
          class="CheckboxList__item"
        >
          <input
            type="checkbox"
            value="1"
          />
          東京都
        </label>
        <label
          class="CheckboxList__item"
        >
          <input
            type="checkbox"
            value="2"
          />
          大阪府
        </label>
      </div>
    </div>
    <div
      class="MainScreen__chart"
    >
      <div
        data-highcharts-chart="0"
        style="overflow: hidden;"
      >
        <div
          class="highcharts-container "
          dir="ltr"
          id="highcharts-0"
          style="position: relative; overflow: hidden; width: 600px; height: 400px; text-align: left; line-height: normal; z-index: 0; user-select: none; outline: none;"
        >
          <svg
            class="highcharts-root"
            height="400"
            style="font-family:\\"Lucida Grande\\", \\"Lucida Sans Unicode\\", Arial, Helvetica, sans-serif;font-size:12px;"
            version="1.1"
            viewBox="0 0 600 400"
            width="600"
            xmlns="http://www.w3.org/2000/svg"
          >
            <desc>
              Created with Highcharts 9.2.2
            </desc>
            <defs>
              <clippath
                id="highcharts-1-"
              >
                <rect
                  fill="none"
                  height="NaN"
                  width="NaN"
                  x="0"
                  y="0"
                />
              </clippath>
              <clippath
                id="highcharts-7-"
              >
                <rect
                  fill="none"
                  height="400"
                  width="NaN"
                  x="-99"
                  y="-10"
                />
              </clippath>
              <clippath
                id="highcharts-8-"
              >
                <rect
                  fill="none"
                  height="NaN"
                  width="NaN"
                  x="0"
                  y="0"
                />
              </clippath>
            </defs>
            <rect
              class="highcharts-background"
              fill="#ffffff"
              height="400"
              rx="0"
              ry="0"
              width="600"
              x="0"
              y="0"
            />
            <rect
              class="highcharts-plot-background"
              fill="none"
              height="NaN"
              width="NaN"
              x="NaN"
              y="10"
            />
            <rect
              class="highcharts-plot-border"
              data-z-index="1"
              fill="none"
              height="375"
              width="580"
              x="10"
              y="10"
            />
            <g
              class="highcharts-grid highcharts-xaxis-grid"
              data-z-index="1"
            >
              <path
                class="highcharts-grid-line"
                d="M 0 0"
                data-z-index="1"
                fill="none"
                opacity="1"
                stroke-dasharray="none"
              />
              <path
                class="highcharts-grid-line"
                d="M 0 0"
                data-z-index="1"
                fill="none"
                opacity="1"
                stroke-dasharray="none"
              />
              <path
                class="highcharts-grid-line"
                d="M 0 0"
                data-z-index="1"
                fill="none"
                opacity="1"
                stroke-dasharray="none"
              />
              <path
                class="highcharts-grid-line"
                d="M 0 0"
                data-z-index="1"
                fill="none"
                opacity="1"
                stroke-dasharray="none"
              />
              <path
                class="highcharts-grid-line"
                d="M 0 0"
                data-z-index="1"
                fill="none"
                opacity="1"
                stroke-dasharray="none"
              />
              <path
                class="highcharts-grid-line"
                d="M 0 0"
                data-z-index="1"
                fill="none"
                opacity="1"
                stroke-dasharray="none"
              />
            </g>
            <g
              class="highcharts-grid highcharts-yaxis-grid"
              data-z-index="1"
            >
              <path
                class="highcharts-grid-line"
                d="M 0 0"
                data-z-index="1"
                fill="none"
                opacity="1"
                stroke="#e6e6e6"
                stroke-dasharray="none"
                stroke-width="1"
              />
              <path
                class="highcharts-grid-line"
                d="M 0 0"
                data-z-index="1"
                fill="none"
                opacity="1"
                stroke="#e6e6e6"
                stroke-dasharray="none"
                stroke-width="1"
              />
              <path
                class="highcharts-grid-line"
                d="M 0 0"
                data-z-index="1"
                fill="none"
                opacity="1"
                stroke="#e6e6e6"
                stroke-dasharray="none"
                stroke-width="1"
              />
              <path
                class="highcharts-grid-line"
                d="M 0 0"
                data-z-index="1"
                fill="none"
                opacity="1"
                stroke="#e6e6e6"
                stroke-dasharray="none"
                stroke-width="1"
              />
              <path
                class="highcharts-grid-line"
                d="M 0 0"
                data-z-index="1"
                fill="none"
                opacity="1"
                stroke="#e6e6e6"
                stroke-dasharray="none"
                stroke-width="1"
              />
              <path
                class="highcharts-grid-line"
                d="M 0 0"
                data-z-index="1"
                fill="none"
                opacity="1"
                stroke="#e6e6e6"
                stroke-dasharray="none"
                stroke-width="1"
              />
              <path
                class="highcharts-grid-line"
                d="M 0 0"
                data-z-index="1"
                fill="none"
                opacity="1"
                stroke="#e6e6e6"
                stroke-dasharray="none"
                stroke-width="1"
              />
            </g>
            <g
              class="highcharts-axis highcharts-xaxis"
              data-z-index="2"
            >
              <text
                class="highcharts-axis-title"
                data-z-index="7"
                style="color:#666666;fill:#666666;"
                text-anchor="end"
                transform="translate(0,0)"
                x="0"
                y="-9999"
              >
                年度
              </text>
              <path
                class="highcharts-axis-line"
                d="M 0 0"
                data-z-index="7"
                fill="none"
                stroke="#ccd6eb"
                stroke-width="1"
              />
            </g>
            <g
              class="highcharts-axis highcharts-yaxis"
              data-z-index="2"
            >
              <text
                class="highcharts-axis-title"
                data-z-index="7"
                style="color:#666666;fill:#666666;"
                text-anchor="end"
                transform="translate(0,0)"
                x="NaN"
                y="0"
              >
                人口数
              </text>
              <path
                class="highcharts-axis-line"
                d="M 0 0"
                data-z-index="7"
                fill="none"
              />
            </g>
            <g
              class="highcharts-series-group"
              data-z-index="3"
            >
              <g
                class="highcharts-series highcharts-series-0 highcharts-line-series highcharts-color-0"
                clip-path="url(#highcharts-8-)"
                data-z-index="0.1"
                opacity="1"
                transform="translate(0,10) scale(1 1)"
              >
                <path
                  class="highcharts-graph"
                  d="M -100000 -100000 L -100000 -100000 L -100000 -100000 L -100000 -100000 L -100000 -100000"
                  data-z-index="1"
                  fill="none"
                  stroke="#7cb5ec"
                  stroke-linecap="round"
                  stroke-linejoin="round"
                  stroke-width="2"
                />
                <path
                  class="highcharts-tracker-line"
                  d="M -100000 -100000 L -100000 -100000 L -100000 -100000 L -100000 -100000 L -100000 -100000"
                  data-z-index="2"
                  fill="none"
                  stroke="rgba(192,192,192,0.002)"
                  stroke-linecap="round"
                  stroke-linejoin="round"
                  stroke-width="22"
                  visibility="visible"
                />
              </g>
              <g
                class="highcharts-markers highcharts-series-0 highcharts-line-series highcharts-color-0 highcharts-tracker"
                clip-path="url(#highcharts-7-)"
                data-z-index="0.1"
                opacity="1"
                transform="translate(0,10) scale(1 1)"
              />
            </g>
            <text
              class="highcharts-title"
              data-z-index="4"
              style="color:#333333;font-size:18px;fill:#333333;"
              text-anchor="middle"
              x="300"
              y="24"
            >
              ㅤ
            </text>
            <text
              class="highcharts-subtitle"
              data-z-index="4"
              style="color:#666666;fill:#666666;"
              text-anchor="middle"
              x="300"
              y="10"
            />
            <text
              class="highcharts-caption"
              data-z-index="4"
              style="color:#666666;fill:#666666;"
              text-anchor="start"
              x="10"
              y="397"
            />
            <g
              class="highcharts-legend highcharts-no-tooltip"
              data-z-index="7"
              transform="translate(590,10)"
            >
              <rect
                class="highcharts-legend-box"
                fill="none"
                height="8"
                rx="0"
                ry="0"
                visibility="visible"
                width="8"
                x="0"
                y="0"
              />
              <g
                data-z-index="1"
              >
                <g>
                  <g
                    class="highcharts-legend-item highcharts-line-series highcharts-color-0 highcharts-series-0"
                    data-z-index="1"
                    transform="translate(8,3)"
                  >
                    <path
                      class="highcharts-graph"
                      d="M 0 11 L 16 11"
                      fill="none"
                      stroke="#7cb5ec"
                      stroke-width="2"
                    />
                    <path
                      class="highcharts-point"
                      d="M 8 15 A 4 4 0 1 1 8.003999999333336 14.999998000000167 Z"
                      fill="#7cb5ec"
                      opacity="1"
                    />
                    <text
                      data-z-index="2"
                      style="color:#333333;cursor:pointer;font-size:12px;font-weight:bold;fill:#333333;"
                      text-anchor="start"
                      x="21"
                      y="15"
                    >
                      東京都
                    </text>
                  </g>
                </g>
              </g>
            </g>
            <g
              class="highcharts-axis-labels highcharts-xaxis-labels"
              data-z-index="7"
            >
              <text
                style="color:#666666;cursor:default;font-size:11px;fill:#666666;"
                text-anchor="middle"
                transform="translate(0,0)"
                x="0"
              >
                1960
              </text>
              <text
                style="color:#666666;cursor:default;font-size:11px;fill:#666666;"
                text-anchor="middle"
                transform="translate(0,0)"
                x="0"
              >
                1965
              </text>
              <text
                style="color:#666666;cursor:default;font-size:11px;fill:#666666;"
                text-anchor="middle"
                transform="translate(0,0)"
                x="0"
              >
                1970
              </text>
              <text
                style="color:#666666;cursor:default;font-size:11px;fill:#666666;"
                text-anchor="middle"
                transform="translate(0,0)"
                x="0"
              >
                1975
              </text>
              <text
                style="color:#666666;cursor:default;font-size:11px;fill:#666666;"
                text-anchor="middle"
                transform="translate(0,0)"
                x="0"
              >
                1980
              </text>
            </g>
            <g
              class="highcharts-axis-labels highcharts-yaxis-labels"
              data-z-index="7"
            >
              <text
                style="color:#666666;cursor:default;font-size:11px;fill:#666666;"
                text-anchor="end"
                transform="translate(0,0)"
                x="0"
              >
                0
              </text>
              <text
                style="color:#666666;cursor:default;font-size:11px;fill:#666666;"
                text-anchor="end"
                transform="translate(0,0)"
                x="0"
              >
                10
              </text>
              <text
                style="color:#666666;cursor:default;font-size:11px;fill:#666666;"
                text-anchor="end"
                transform="translate(0,0)"
                x="0"
              >
                20
              </text>
              <text
                style="color:#666666;cursor:default;font-size:11px;fill:#666666;"
                text-anchor="end"
                transform="translate(0,0)"
                x="0"
              >
                30
              </text>
              <text
                style="color:#666666;cursor:default;font-size:11px;fill:#666666;"
                text-anchor="end"
                transform="translate(0,0)"
                x="0"
              >
                40
              </text>
              <text
                style="color:#666666;cursor:default;font-size:11px;fill:#666666;"
                text-anchor="end"
                transform="translate(0,0)"
                x="0"
              >
                50
              </text>
              <text
                style="color:#666666;cursor:default;font-size:11px;fill:#666666;"
                text-anchor="end"
                transform="translate(0,0)"
                x="0"
              >
                60
              </text>
            </g>
            <text
              class="highcharts-credits"
              data-z-index="8"
              style="cursor:pointer;color:#999999;font-size:9px;fill:#999999;"
              text-anchor="end"
              x="590"
              y="395"
            >
              Highcharts.com
            </text>
          </svg>
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
